File: labs/Module_2_JAVA_LAB_1/Reverse_A_String.java
import java.util.Scanner;
// change the class name to Main
public class Reverse_A_String {
    public static String reversePreserve(String text) {
        StringBuilder letters = new StringBuilder();
        // Collect all alphabetic characters
        for (char c : text.toCharArray()) {
            if (Character.isLetter(c)) {
                letters.append(c);
            }
        }
        
        // Reverse the collected characters
        letters.reverse();
        
        // Reintegrate characters
        StringBuilder result = new StringBuilder();
        int letterIndex = 0;
        for (char c : text.toCharArray()) {
            if (Character.isLetter(c)) {
                result.append(letters.charAt(letterIndex++));
            } else {
                result.append(c);
            }
        }
        
        return result.toString();
    }
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        // Prompt the user for input
        // System.out.println("Enter a string to be reversed, preserving punctuation and spaces:");
        String userInput = scanner.nextLine();
        
        // Process the input and display the output
        String output = reversePreserve(userInput);
        // System.out.println("Reversed string: " + output);
        System.out.println(output);
        
        scanner.close();
    }
}
                                                                                                                            

File: labs/Module_2_Python_LAB_1/Reverse_a_string.py

def reverse_string_preserving_special_characters(input_string):
    # Extract letters and store non-letter positions with their characters
    letters = [char for char in input_string if char.isalpha()]
    letters.reverse()  # Reverse the list of letters

    # Reconstruct the string with reversed letters in the correct positions
    output_list = list(input_string)  # Convert string to list to modify it
    letter_index = 0  # Start with the first letter in the reversed list

    for i in range(len(output_list)):
        if output_list[i].isalpha():  # If the character is a letter
            output_list[i] = letters[letter_index]  # Replace it with the reversed letter
            letter_index += 1  # Move to the next letter in the reversed list

    return ''.join(output_list)  # Join list back into string

def main():
    # Get user input
    # print("Enter the string to be reversed (punctuation and spaces will be preserved):")
    user_input = input()
    
    # Process and output the reversed string
    result = reverse_string_preserving_special_characters(user_input)
    # print("Reversed string:", result)
    print(result)

if __name__ == "__main__":
    main()